src/mgr/ServiceMap.cc

```diff
@@ -75,20 +77,99 @@ std::string ServiceMap::Service::get_summary() const
   if (daemons.empty()) {
     return "no daemons active";
   }
+
   std::ostringstream ss;
-  ss << daemons.size() << (daemons.size() > 1 ? " daemons" : " daemon")
-     << " active";
 
-  if (!daemons.empty()) {
-    ss << " (";
+  // The format is one pair in metadata:
+  //   "classify" : "${type}/${prefix}"
+  //
+  // type: will be used to replace the default "daemon(s)"
+  // showed in `ceph -s`. If absent, the "daemon" will be used.
+  // prefix: if present the active members will be classified
+  // by the prefix instead of "daemon_name".
+  //
+  // For exmaple for iscsi gateways, it will be something likes:
+  //   "classify" + "\0" + "portal/gateway0" + "\0"
+  // The `ceph -s` will be something likes:
+  //    iscsi: 3 portals active (gateway0, gateway1, gateway2)
+
+  std::set<std::string> prefs;
+  bool classified = false;
+  auto p = daemons.begin();
+  auto q = p->second.metadata.find("classify");
+  std::string first_type;
+  if (q != p->second.metadata.end()) {
+    const std::regex pattern("(\\w*)/(\\w*)");
+    std::smatch match;
+    if (std::regex_search(q->second, match, pattern)) {
+      first_type = std::string(match[1]);
+      std::string first_prefix = std::string(match[2]);
+      if (first_type.empty() || first_prefix.empty()) {
+        // Invalid pattern, just show what they are one by one
+        classified = false;
+      } else {
+        classified = true;
+        prefs.insert(first_prefix);
+
+        while (++p != daemons.end()) {
+          q = p->second.metadata.find("classify");
+          if (q == p->second.metadata.end()) {
+            // No classify metadata found, just show what they are one by one
+            classified = false;
+            break;
+          }
+
+          if (!std::regex_search(q->second, match, pattern)) {
+             // The key patterns are inconsistent for all the daemons,
```

Comment:
This seems like an edge case. Instead of worrying about some daemons using the new format and some using old format (w/o classification support), I'd suggest instead just build a map from daemon type to matching daemons first (where that special "daemon_type" metadata key is evaluated).

Reply:
Sounds perfect, something like:

`std::map<<std::string, std::set<std::string>> prefs;`